File: bench/ocaml/binary_trees.ml
(* Binary Trees benchmark — matches bench/binary_trees.march (depth=15) *)
type tree = Leaf | Node of tree * tree

let rec make d = if d = 0 then Leaf else Node (make (d - 1), make (d - 1))
let rec check = function Leaf -> 1 | Node (l, r) -> check l + check r + 1

let () =
  let n = 15 and min_depth = 4 in
  let max_depth = max n (min_depth + 2) in
  let stretch = max_depth + 1 in
  Printf.printf "stretch tree of depth %d check: %d\n" stretch (check (make stretch));
  let long_lived = make max_depth in
  let d = ref min_depth in
  while !d <= max_depth do
    let iters = 1 lsl (max_depth - !d + min_depth) in
    let sum = ref 0 in
    for _ = 1 to iters do sum := !sum + check (make !d) done;
    Printf.printf "%d trees of depth %d check: %d\n" iters !d !sum;
    d := !d + 2
  done;
  Printf.printf "long lived tree of depth %d check: %d\n" max_depth (check long_lived)

File: bench/rust/binary_trees.rs
// Binary Trees benchmark — matches bench/binary_trees.march (depth=15)
enum Tree { Leaf, Node(Box<Tree>, Box<Tree>) }

fn make(d: u32) -> Tree {
    if d == 0 { Tree::Leaf } else { Tree::Node(Box::new(make(d - 1)), Box::new(make(d - 1))) }
}

fn check(t: &Tree) -> i64 {
    match t {
        Tree::Leaf => 1,
        Tree::Node(l, r) => check(l) + check(r) + 1,
    }
}

fn main() {
    let (n, min_depth): (u32, u32) = (15, 4);
    let max_depth = n.max(min_depth + 2);
    let stretch = max_depth + 1;
    println!("stretch tree of depth {} check: {}", stretch, check(&make(stretch)));

    let long_lived = make(max_depth);
    let mut d = min_depth;
    while d <= max_depth {
        let iters = 1i64 << (max_depth - d + min_depth);
        let sum: i64 = (0..iters).map(|_| check(&make(d))).sum();
        println!("{} trees of depth {} check: {}", iters, d, sum);
        d += 2;
    }
    println!("long lived tree of depth {} check: {}", max_depth, check(&long_lived));
}
